### PR TITLE
Refactored PatternScanner to allow for the caller to specify an offse…

### PR DESCRIPTION
**Pull request by gir489**
…t from the module's BaseAddress to start the search from.

### CODE REVIEW

File: src/Process.NET/Patterns/PatternScanner.cs [hunk 1]
```diff
@@ -7,11 +7,18 @@ namespace Process.NET.Patterns
     public class PatternScanner : IPatternScanner
     {
         private readonly IProcessModule _module;
+		private readonly int _offsetFromBaseAddress;
```
Comment (gir489) [inline src/Process.NET/Patterns/PatternScanner.cs]: Fucking tabs.

**Comment by lolp1** [comment]
@gir489 I am a little concerned with this line. Perhaps I am confused, clarify if so :)

https://github.com/gir489/Process.NET/blob/62bb19080d04fde5aeef610080fa4e25dad04640/src/Process.NET/Patterns/PatternScanner.cs#L77

The idea behind the read with offset + pattern.Offset there is because often you need to find a location in memory to start from, and then resolve what is at that location + some specified bytes extra. I'm not sure how that extra offset from the module's base address is relevant there,. 

Imagine:
```
var scanResult = (...) // finds a memory location that I can resolve my desired address from
var realAddress = (IntPtr)memory.read<byte>(scanResult.BaseAddress);
// or 
var realAddress = memory.read<IntPtr>(scanResult.BaseAddress + offset);

**Comment by gir489** [comment]
@lolp1 I saw that while looking at the code after I committed. I was actually working on a commit as I saw you made a comment. I didn't know what the function did at the time, until I read it a little more. I guess it's for retrieving addresses at a certian signature? For example if the signature was "48 8D 05 ? ? ? ? 48 89 CF", it would return the value of the 4 question marks in the RealAddress field?

**Comment by lolp1** [comment]
You're about right. The goal of a pattern scan is to find a unique location in a module's memory. To do this, a mask and wild card system is used. The simplest patterns to find are static function pointers.. . 

E.g this would be an example pattern for a static function.  "48 89 74 24 ?? 57 48 83 EC 20 48 8B 05 ?? ?? ?? ?? 48 8B F1 48 8B FA 48 8B 88 ?? ?? ?? ?? F6 C1 01 75 05 48 85 C9 75 02".. the offset this pattern scan results in just needs to be rebased.

The second method is to use the same pattern matching technique, but this time it is used to find a _starting_ point. This is because the address you want is not at a static location itself (unlike a static function), _but_ a static location _accesses_ the address. So in those cases, we need to find one of those static locations accessing the address we want with a pattern.

E.g this finds an offset, our start: "85 c0 0f 84 ?? ?? ?? ?? 8b 06 8b 50 ?? 8d 4d ?? 51 8b ce ff d2 d9 45 ?? 8b 46 ??". Now we need to resolve the address we want, which we know is at module base + pattern offset found + 1 byte. So we read it...

```
var realAddress = (IntPtr)memory.Read<byte>(patternScanResult.BaseAddress);
```

**Comment by gir489** [comment]
You're talking about the JMP math x86 does on CALL instructions? That's just instruction pointer + value.

The idea being, instead of doing something logical like MOV EIP, VALUE. Intel decided to do ADD EIP, VALUE, because reasons. This presents a problem when trying to convert the value to an address relative to the module on an above bit platform, because it expects the addition to drop the overflowed 1 on the left, and end up with the address. I've never had to figure out a CALL address externally, although I don't know why I really would, because I'd have no ACTUAL way of making a call to it... I'd have to make some sort've code cave to call it.

**Comment by lolp1** [comment]
Here is an example for clarity.

```
// a simple static function offset
addr=FindBinary(0, SEARCH_DOWN,"48 89 5C 24 ? 57 48 83 EC 20 48 8B 3D ? ? ? ? 48 8B D9 45 33 C0 48 8B CF E8 ? ? ? ?"); 
fprintf( hFile," \tinternal static readonly IntPtr %s = ImageBase + 0x%08X;\n", name, addr-0x140000000);

// finding an offset that is read at player address + offse.
addr=FindBinary(0, SEARCH_DOWN,"F7 83 ? ? ? ? ? ? ? ? 74 40 49 8B CC E8 ? ? ? ? 4C 8D 05 ? ? ? ? 48 8D 0D ? ? ? ? ");
b = Byte(addr +2);
bb = Byte(addr + 3);
fprintf( hFile," \tinternal static readonly int %s = 0x%X%X;\n", name, bb,b);

// finding a data pointer address
addr=FindBinary(0, SEARCH_DOWN,"48 8B 1D ? ? ? ? 44 8B 15 ? ? ? ? 44 8B C8 41 F7 D9 8B D0 85 D2");
b = Dword(addr +3);
addr = (addr +7) + b ;
fprintf( hFile," \t IntPtr %s = ImageBase + 0x%X;\n", name,addr- 0x140000000);
```

**Comment by gir489** [comment]
I understand what it does now, but I don't think DwordPattern itself should enable that behavior just by overloading the constructor. I think you should have something like DwordDataPattern, to imply that it's going to use a different search behavior.

**Comment by lolp1** [comment]
@gir489 I agree. I want to implement it more like I see here: https://github.com/Vipeax/cleanLayer/tree/c59e26bf227c1a91c3e63c1fc6478a4d15151b26/cleanPattern

Feel free to do so and pull req if you would like ^_^ I think the current one is just kind of patched together and needs many improvements, despite it "working".

**Comment by gir489** [comment]
I'll make a separate pull request for that. I actually had some ideas to implement other various forms of the MemoryPatternType, such as CALL, LJMP, and String. Those are the 3 most common things I've had to signature find before. [https://bitbucket.org/gir489/m0d-s0beit-v-redux/src/2b28c2c2e9089d872f2d048a9f564c82f12401f3/m0d-s0beit-v/RAGEHelper.cpp?at=master&fileviewer=file-view-default#RAGEHelper.cpp-166](I've had to just cast the ASCII values to byte form in the past for String, when it could easily be done by the compiler)

**Comment by lolp1** [comment]
Glad to see a CPP user take interest :) I will try and get the latest git/nuget out tomorrow afternoon. Yeah some patterns are a pain.. like having to chain pattern scan results just to find a unique pattern.

**Comment by gir489** [comment]
This pull request is ready for review and merging with the project. I've thoroughly tested it.

**Comment by gir489** [comment]
I guess I can't make a separate pull request for the Boyer-Moore-Horspool, it got mashed in to this one...

**Comment by lolp1** [comment]
Why the switch to "Boyer-Moore-Horspool"? Benchmark shows at least on my PC a performance _decrease_

**Comment by gir489** [comment]
When I benched it on RE7 for a pattern of "0F 85 ? ? ? ? 45 89 F1 C6 44 24" starting at BaseAddress + 0x73B1000, it took 1,337 milliseconds to do the Naive method, and 31 seconds for the Boyer-Moore-Horspool.

Boyer-Moore-Horspool will always be more efficient, if not the same when executed properly, because in essence, it's starting out doing a naive search, then instead of +1ing the index, it has a table it uses to skip by. If you'd like to know more about it, watch this video: https://www.youtube.com/watch?v=PHXAOKQk2dw

**Comment by lolp1** [comment]
Just as a side not what is the need for such performance? Pattern scans  happen at most one time on app start up?

**Comment by gir489** [comment]
Those savings may seem menial now, but they stack up extreemly fast, espeically in something ([like what I'm using it for](http://forum.cheatengine.org/viewtopic.php?t=602920)) that requires the signatures to finish before finishing execution. Obviously, I can only thing of cheats that this would apply to, since you don't want to be listening to my keygen music for very long once the app starts. Originally the trainer took 20 seconds before the music stopped. That time was dropped to 3 seconds with my implementation of offsetFromBaseAddress. That time was then dropped to <1 second with my BMH implementation. It almost seems fluid for me. The music stops at the exact moment RE7 appears to open, despite the RE7 having been running for several seconds, 1 of which I have to sit there waiting for the app to decrypt itself.

At any rate, I just committed an NUnit which you can use to benchmark your Naive algorithm, and my implementation of Boyer-Moore-Horspool. Pic: (http://www.mediafire.com/convkey/67be/rpt5zej966tj98uzg.jpg)

You can see, the BMH algorithm destories it in terms of runtime, and that's just because of the nature of BMH. The runtimes savings scale exponentially the longer the patter you're trying to match. I'm surprised it got such a massive improvement on such a common pattern that was at 0x0040105F in IDA.

**Comment by gir489** [comment]
It seems there was a bug with my reimplementation of your Naive search, and now I get the same results as you.

My guess is that the extra time it takes for BMH to fill the bad shift table isn't ending up being "faster" by much if the pattern is extreemly close to the BaseAddress. However, the difference is negligible, compared to the potential savings of a long search, and why I included a method to switch between the two.

This can been seen to its full potential when the BMH failure pattern finished exponentially faster than Naive, because of the longer pattern, and that it's searching the entire module.

**Comment by gir489** [comment]
After further testing, it seems Naive search is super optimized for 32-bit, considering it doesn't need to take the extra 10 milliseconds to construct the BMH table, and for 64-bit it's almost in terms of exponential.

Here are the 4 tests I originally did in just inline C#, now ported to NUnit, to show side-by-side the runtime comparisons:  http://www.mediafire.com/view/4y937n39j3hkcl7/BMHRE7.png

As you can see, without using the offset and using Naive, it took about 10 seconds to run the test, which is the results I was getting before. That's ONE pattern! I think for 64-bit, it should default to BMH, and for 32-bit use Naive, since it does seem to be objectively faster.

**Comment by gir489** [comment]
You still haven't merged this? Why?

**Comment by lolp1** [comment]
@gir489 

Done, sorry.

**Pull request merged**